src/DataService/CachingContextFactory.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+namespace Lullabot\Mpx\DataService;
+
+use phpDocumentor\Reflection\Types\Context;
+
+/**
+ * Creates a context for docblocks, storing cached results for already parsed files.
+ *
+ * @see \phpDocumentor\Reflection\Types\ContextFactory
+ */
+final class CachingContextFactory
+{
+    /** The literal used at the end of a use statement. */
+    const T_LITERAL_END_OF_USE = ';';
+
+    /** The literal used between sets of use statements */
+    const T_LITERAL_USE_SEPARATOR = ',';
+
+    /**
+     * Build a Context given a Class Reflection.
+     *
+     * @param \Reflector $reflector
+     *
+     * @see Context for more information on Contexts.
+     *
+     * @return Context
+     */
+    public function createFromReflector(\Reflector $reflector)
+    {
+        if (method_exists($reflector, 'getDeclaringClass')) {
+            $reflector = $reflector->getDeclaringClass();
+        }
+
+        $fileName = $reflector->getFileName();
+        $namespace = $reflector->getNamespaceName();
+
+        if (file_exists($fileName)) {
+            return $this->createForNamespace($namespace, file_get_contents($fileName));
+        }
+
+        return new Context($namespace, []);
+    }
+
+    /**
+     * Build a Context for a namespace in the provided file contents.
+     *
+     * @param string $namespace    It does not matter if a `\` precedes the namespace name, this method first normalizes.
+     * @param string $fileContents the file's contents to retrieve the aliases from with the given namespace.
+     *
+     * @see Context for more information on Contexts.
+     *
+     * @return Context
+     */
+    public function createForNamespace($namespace, $fileContents)
+    {
+        static $cache = [];
```

Comment:
This is the reason I had to copy this class. The details?

I had to switch from using the ReflectionExtractor to the PhpDocExtractor. That means that Symfony discovers what types of objects a property accepts from the phpdocs instead of the phpcode. Unfortunately, PHP itself doesn't have generics (the C# / Java term) meaning "this property accepts an array of Ratings". PHP code can only indicate that it at accepts an array, but not what type each array item must be.

The PHPDoc extractor lets us write something like `\DateTime[]` and Symfony will know that all set calls for that property must be an array of \DateTime objects.

If you read this function, you'll see it re-parses `$fileContents` every time it's called. So, for a MediaFile with tons of properties, it's hundreds of times which is totally not OK for a first-time load. Since this method knows that `$fileContents` will always parse identically, we can do a static cache off of it's value. Before this cache, it was a huge performance hit here - 45-90 seconds!

Why a basic static variable, and not a proper cache backend? We are already wrapping this in a cache backend, but that cache set only happens after a class has been parsed completely. It doesn't incrementally set the cache as it goes.